plugin/hosts/hostsfile.go

```diff
@@ -52,71 +44,79 @@ type Hostsfile struct {
 	// including IPv6 address with zone identifier.
 	// We don't support old-classful IP address notation.
 	byAddr map[string][]string
+}
 
-	// inline saves the hosts file is inlined in Corefile
-	// We need a copy here as we want to use inline to override
-	// the default /etc/hosts
-	inline []string
+func newHostsMap() *hostsMap {
+	return &hostsMap{
+		byNameV4: make(map[string][]net.IP),
+		byNameV6: make(map[string][]net.IP),
+		byAddr:   make(map[string][]string),
+	}
+}
 
-	expire time.Time
-	path   string
-	mtime  time.Time
-	size   int64
+// Hostsfile contains known host entries.
+type Hostsfile struct {
+	sync.RWMutex
+
+	// list of zones we are authoritive for
+	Origins []string
+
+	// hosts maps for lookups
+	hmap *hostsMap
+
+	// inline saves the hosts file that is inlined in a Corefile.
+	// We need a copy here as we want to use it to initialize the maps for parse.
+	inline *hostsMap
+
+	// path to the hosts file
+	path string
+
+	// mtime and size are only read and modified by a single goroutine
+	mtime time.Time
+	size  int64
 }
 
 // ReadHosts determines if the cached data needs to be updated based on the size and modification time of the hostsfile.
 func (h *Hostsfile) ReadHosts() {
-	now := time.Now()
-
-	// Not deferring h.RUnlock() because we may need to remove the read lock and aquire a write lock
-	h.RLock()
-	if now.Before(h.expire) && len(h.byAddr) > 0 {
-		h.RUnlock()
+	file, err := os.Open(h.path)
+	if err != nil {
```

Comment:
This might benefit from a short comment that we don't care if the file does not exist (hence we're not returning `err`)